### PR TITLE
LibTLS+LibCrypto: Implement signature checking for DHE and ECDHE

### PR DESCRIPTION
**Pull request by msvisser**
This pull request implements the signature checking for the DHE and ECDHE key exchanges. Implementing this requires the EMSA-PKCS1-V1_5 encoding, which is also implemented in this pull request.

It also contains some error cleanup in LibTLS, which was originally introduced in #12625.

> This pull request makes sure that an OOM condition will actually result in an error in the DHE and ECDHE key exchange instead of simply returning no error and hoping it will break elsewhere.
> 
> It also changes some of the FeatureNotSupported errors in ECDHE to NotUnderstood, as a compliant server should never send values that will result in the errors. Furthermore, NotUnderstood will actually generate an InternalError TLS alert, instead of crashing the RequestServer.

### CODE REVIEW

File: Userland/Libraries/LibTLS/HandshakeServer.cpp [hunk 1]
```diff
@@ -317,9 +318,82 @@ ssize_t TLSv12::handle_ecdhe_rsa_server_key_exchange(ReadonlyBytes buffer)
...
+    }
+    auto message = message_result.release_value();
+    message.overwrite(0, m_context.local_random, 32);
+    message.overwrite(32, m_context.remote_random, 32);
+    message.overwrite(64, server_key_info_buffer.data(), server_key_info_buffer.size());
+
+    Crypto::VerificationConsistency verification;
+    switch ((HashAlgorithm)signature_hash) {
+    case HashAlgorithm::SHA1: {
+        auto pkcs1 = Crypto::PK::EMSA_PKCS1_V1_5<Crypto::Hash::SHA1>();
```
Comment (alimpfard) [inline Userland/Libraries/LibTLS/HandshakeServer.cpp:368]: You can probably condense these with a Hash::Manager, if your emsa pkcs#1 v1.5 ctor passes arguements to the hash object ctor.

**Review by alimpfard — COMMENTED** [review]
Looks generally fine :)

File: Userland/Libraries/LibTLS/HandshakeServer.cpp (hunk 1 above)
Comment (msvisser) [reply Userland/Libraries/LibTLS/HandshakeServer.cpp]: I'm not sure if the `Hash::Manager` implementation is much better, it still requires the case to translate the `TLS::HashAlgorithm` to `Crypto::Hash::HashKind`. It also requires extra logic in `Crypto::PK::EMSA_PKCS1_V1_5` to generate the correct digest info based on the type of hash used by the hash manager.

Comment (alimpfard) [reply Userland/Libraries/LibTLS/HandshakeServer.cpp]: Ah yeah, I missed the extra data it needs, it looks fine to me though, so :shrug:

Comment (msvisser) [reply Userland/Libraries/LibTLS/HandshakeServer.cpp]: Yeah it's fine like this

**Pull request merged**